core/templates/dev/head/services/AssetsBackendApiService.ts

```diff
@@ -33,29 +33,47 @@ oppia.factory('AssetsBackendApiService', [
       $http, $q, AudioFileObjectFactory, CsrfTokenService,
       FileDownloadRequestObjectFactory, ImageFileObjectFactory,
       UrlInterpolationService, DEV_MODE) {
-    if (!DEV_MODE && !GLOBALS.GCS_RESOURCE_BUCKET_NAME) {
-      throw Error('GCS_RESOURCE_BUCKET_NAME is not set in prod.');
-    }
+    var GCS_RESOURCE_BUCKET_NAME = null;
+    var ASSET_TYPE_AUDIO = 'audio';
+    var ASSET_TYPE_IMAGE = 'image';
+    var GCS_PREFIX = null;
+    var AUDIO_DOWNLOAD_URL_TEMPLATE = null;
+    var IMAGE_DOWNLOAD_URL_TEMPLATE = null;
 
     // List of filenames that have been requested for but have
     // yet to return a response.
     var _audioFilesCurrentlyBeingRequested = [];
     var _imageFilesCurrentlyBeingRequested = [];
+    var AUDIO_UPLOAD_URL_TEMPLATE =
+      '/createhandler/audioupload/<exploration_id>';
 
-    var ASSET_TYPE_AUDIO = 'audio';
-    var ASSET_TYPE_IMAGE = 'image';
+    var _initialize = function() {
+      if (DEV_MODE) {
+        AUDIO_DOWNLOAD_URL_TEMPLATE = (
```

Comment:
I looked at the file more carefully and I think I understand what the issue is. Now that we are fetching the bucket name asynchronously, you need to wait for that promise to resolve before you can actually do anything with regards to the assets.

I think the current approach is fraught, though. No developer is going to remember to call initialize() explicitly and there aren't really any rules for when it should be called. So this is not really maintainable, and that's why, typically, a service takes care of it's own initialization. The current approach is a bit scattershot and relies on other files telling this service when to initialize.

Instead, this service should still take care of its own initialization, but calls to this service should still always get the correct responses. Can you find a way to do this such that, if a (presumably async) call to serve stuff comes in when the GCS bucket data hasn't come in yet, that call gets served only when that data has come in? I think I saw a recent PR by you or @brianrodri that did something like this, but I can't remember which one it was -- IIRC it involved storing a promise in a local variable in the service and returning that.

Reply:
I have refactored it to initialize itself and the other methods returns a promise which is fulfilled after the bucket name has been fetched.
PTAL.